scientific-computing-with-python/learn-encapsulation-by-drawing-a-projectile-trajectory/projectile-trajectory.py

```diff
@@ -0,0 +1,154 @@
+            self.__speed * math.sin(self.__angle) +
+            math.sqrt(self.__speed**2 * math.sin(self.__angle)**2 +
+                      2 * GRAVITATIONAL_ACCELERATION * self.__height)) / GRAVITATIONAL_ACCELERATION
+
+    def __calculate_y_coordinate(self, x):
+        #TODO: break in smaller pieces
+        return self.__height + math.tan(self.__angle) * x - GRAVITATIONAL_ACCELERATION * x**2 / (
+            2 * self.__speed**2 * math.cos(self.__angle)**2)
+
+    def __calculate_all_coordinates(self):
+        return [
+            (x, self.__calculate_y_coordinate(x))
+            for x in range(math.ceil(self.__displacement))
+        ]
+
+    def __create_coordinates_table(self):
+        table = '  x      y\n'
+        for x, y in self.__coordinates:
+            table += f'{x:>3}{round(y, 2):>7.2f}\n'
+
+        return table
+
+    def __create_trajectory(self):
+        rounded_coord = [(x, round(y)) for x, y in self.__coordinates]
+        rows = max([y for x, y in rounded_coord]) + 1
+        columns = max([x for x, y in rounded_coord]) + 1
+        # TODO?: if rows and/or columns too big, no draw (how big?)
+        graph = ''
+        for y in range(rows, -3, -1):
+            row = ''
+            for x in range(-2, columns + 1):
+                if (x, y) in rounded_coord:
+                    row += PROJECTILE
+                elif x == -1 and y >= 0:
+                    row += '⊣'
+                elif y == -1 and x >= 0:
+                    row += '⊤'
+                elif (x == -2 and y == 0) or (x == 0 and y == -2):
+                    row += '0'
+                else:
+                    row += ' '
+            graph += row + '\n'
+        return graph
+
+    def __str__(self):
+        return self.__create_trajectory()
+
+    def table(self):
+        return self.__create_coordinates_table()
+
+    def details(self):
+        return f'angle: {round(math.degrees(self.__angle))}°\nspeed: {self.__speed} m/s\nstarting height: {self.__height} m\ndisplacement: {round(self.__displacement, 1)} m'
+
+    @property
+    def height(self):
+        return self.__height
+    
+    @height.setter
+    def height(self, new_height):
+        self.__height = new_height
+        self.__displacement = self.__calculate_displacement()
+        self.__coordinates = self.__calculate_all_coordinates()
+
+    @property
+    def angle(self):
+        return self.__angle
+    
+    @angle.setter
+    def angle(self, new_angle):
+        self.__angle = math.radians(new_angle)
+        self.__displacement = self.__calculate_displacement()
+        self.__coordinates = self.__calculate_all_coordinates()
+
+    @property
+    def speed(self):
+        return self.__speed
+
+    @speed.setter
+    def speed(self, new_speed):
+        self.__speed = new_speed
+        self.__displacement = self.__calculate_displacement()
+        self.__coordinates = self.__calculate_all_coordinates()
+
+def terminal_menu():
```

Comment:
I will later take another look at this function, there's a bit that can be done to make it simpler and easier for potential extending.